### PR TITLE
ARROW-8702: [Packaging][C#] Build NuGet packages in release process

### PR DESCRIPTION
**Pull request by kou**
We released NuGet packages as unofficial packages because we only vote
Apache Arrow C# source code. We didn't vote NuGet packages.

We can vote NuGet packages with this change. We can mark NuGet
packages as official packages in the next release.

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit nuget

File: dev/tasks/crossbow.py [hunk 1]
```diff
@@ -706,7 +706,25 @@ def get_version(root, **kwargs):
...
+    # > arbitrary string immediately following the patch version and a
+    # > dash. The string MUST be comprised of only alphanumerics plus
+    # > dash [0-9A-Za-z-].
+    #
+    # Example:
+    #
+    #   '0.16.1.dev10' ->
+    #   '0.16.1-dev10'
+    next_version = re.sub(r'\.(dev\d+)$', r'-\1', next_version)
+    return next_version
```
Comment (kou) [inline dev/tasks/crossbow.py:727]: @kszucs What do you think about this? Is it acceptable?

Comment (kszucs) [reply dev/tasks/crossbow.py:727]: Looks fine for me, but we need to test whether the wheels and conda pkgs accept this format.

**Comment by nealrichardson** [comment]
For what it's worth, did anyone even verify the C# source in the release? I know I couldn't get it to run, and I seemed to recall others saying they skipped it. That is to say, we should make sure our release verification process keeps up if we're adding more release artifacts.

File: dev/tasks/crossbow.py (hunk 1 above)
Comment (kou) [reply dev/tasks/crossbow.py:727]: OK. I'll run jobs for them.

**Comment by kou** [comment]
@github-actions crossbow submit -g conda -g wheel

**Comment by kou** [comment]
@github-actions crossbow submit nuget

**Comment by kou** [comment]
@github-actions crossbow submit nuget

**Comment by kou** [comment]
@github-actions crossbow submit -g conda -g wheel

**Comment by kou** [comment]
@nealrichardson I always verified C# source in each release. The verification always passes tests but failed `sourcelink test`. So I always report "I skipped C# source verification".

The `sourcelink test` failure has been fixed in #7040. We will not need to skip C# source verification in the next release.

**Comment by kou** [comment]
@eerhardt Could you confirm whether NuGet packages at https://github.com/ursa-labs/crossbow/releases/tag/actions-205-github-nuget is valid?

**Review by eerhardt — COMMENTED** [review]
> @eerhardt Could you confirm whether NuGet packages at https://github.com/ursa-labs/crossbow/releases/tag/actions-205-github-nuget is valid?

I gave it a smoke test run and it worked for me. So I believe it is valid.

File: dev/tasks/crossbow.py (hunk 1 above)
Comment (kou) [reply dev/tasks/crossbow.py:727]: Umm... Conda doesn't like it:

```text
Error: bad character '-' in package/version: 0.17.1-dev124
```

File: docker-compose.yml [hunk 2]
```diff
@@ -1015,7 +1015,8 @@ services:
     command: &csharp-command >
       /bin/bash -c "
         /arrow/ci/scripts/csharp_build.sh /arrow &&
-        /arrow/ci/scripts/csharp_test.sh /arrow"
+        /arrow/ci/scripts/csharp_test.sh /arrow &&
+        /arrow/ci/scripts/csharp_nuget.sh /arrow"
```
[the remark this answers is not in the document]
Comment (kou) [reply docker-compose.yml]: It makes sense! Done.

File: ci/scripts/csharp_nuget.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+# under the License.
+
+set -eux
+
+source_dir=${1}/csharp
+
+pushd ${source_dir}
+dotnet pack -c Release
+for pdb in artifacts/Apache.Arrow/Release/*/Apache.Arrow.pdb; do
+  sourcelink test ${pdb}
```
[the remark this answers is not in the document]
Comment (kou) [reply ci/scripts/csharp_nuget.sh]: I agree with you.
Done.

**Comment by kou** [comment]
@eerhardt Thanks for confirming the packages!

**Comment by kou** [comment]
@github-actions crossbow submit nuget -g conda -g wheel

File: dev/tasks/crossbow.py (hunk 1 above)
Comment (kou) [reply dev/tasks/crossbow.py]: I've added `arrow.no_rc_semver_version` instead of changing the current `arrow.version`.

**Comment by kou** [comment]
@github-actions crossbow submit nuget

**Comment by kou** [comment]
@github-actions crossbow submit nuget

**Review by eerhardt — APPROVED** [review]
The csharp parts look good to me.

**Comment by kou** [comment]
I found a problem that URLs embedded in `.pdb` use fork repository (`kou/arrow` in this case). It's required for general cases. If we always use `apache/arrow` here, CI will be failure for commits that changes C# sources. But it must use `apache/arrow` in release process.

I'll work on this as a follow-up task.

I'll merge this.

**Pull request closed without merging**